### PR TITLE
RHDEVDOCS-3924 - document option to add cluster id to off-cluster int…

### PR DESCRIPTION
**Pull request by bburt-rh**
Summary: This PR documents how to add cluster ID labels to metrics being sent to remote write storage.

- Aligned team: DevTools
- For branches: 4.11+
- Jira: https://issues.redhat.com/browse/RHDEVDOCS-3924
- Direct link to doc preview: http://file.rdu.redhat.com/bburt/RHDEVDOCS-3924-option-to-add-cluster-id-to-off-cluster-integrations/monitoring/configuring-the-monitoring-stack.html#adding-cluster-id-labels-to-metrics_configuring-the-monitoring-stack
- SME review: @jan--f 
- QE review: @lihongyan1 
- Peer review: @jc-berger

### CODE REVIEW

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        writeRelabelConfigs:
+        - sourceLabels: [__name__]
+          regex: 'my_metric'
+          action: keep
+        - sourceLabels:
+            - __tmp_openshift_cluster_id__ <1>
+          targetLabel: cluster_id_1 <2>
+          action: replace <3>
+----
+<1> The name of a temporary source label for the cluster ID label. This label must begin with [x-]`__tmp_openshift_` followed by the cluster ID label you want to add followed by [x-]`__` as indicated in the sample.
```
Comment (jan--f) [inline modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:90]: Small correction: The source label is always going to be `__tmp_openshift_cluster_id__` for this cluster id feature. More details for context, this doesn't need to be part of the docs:
CMO adds this temporary label by adding two relabel steps to the `writeRelabelConfigs` of all `remoteWrite` sections. It adds an entry as the first step to add the label called `__tmp_openshift_cluster_id__` with the value of the cluster id. It also adds a step to the end of the `writeRelabelConfigs` removing the label called `__tmp_openshift_cluster_id__`.
A user can add a relabel step as documented here to take the value of this temporary label and add it to their own (non-temporary label). They can choose the name of that label as they see fit.

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+<2> The name of the cluster ID label added to metrics sent to remote write storage. The first part of this value must match the temporary value followed by a unique identifier for this cluster.
```
Comment (jan--f) [inline modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:91]: Like outlined above user are completely free to choose this label name. It might be worth adding some notes about `relabelConfigs` in general:
- If a user chooses a label name that already exists on a metric, the value is overwritten with the cluster id.
- A user should not attempt to use `__tmp_openshift_cluster_id__` as their chosen label name, as this label will be removed as the final relabeling step.

**Review by jan--f — CHANGES_REQUESTED** [review]

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc (hunk 1 above)
Comment (bburt-rh) [reply modules/monitoring-creating-cluster-id-labels-for-metrics.adoc]: Revised per your explanation.

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc (hunk 2 above)
Comment (bburt-rh) [reply modules/monitoring-creating-cluster-id-labels-for-metrics.adoc]: Revised per your suggestion

**Comment by bburt-rh** [comment]
@lihongyan1 Please review and provide feedback. Thank you!

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+metadata:
+  name: cluster-monitoring-config
+  namespace: openshift-monitoring
+data:
+  config.yaml: |
+    prometheusK8s:
+      remoteWrite:
+      - url: "https://remote-write-endpoint.example.com"
+        <endpoint_authentication_credentials>
+        <write_relabel_configs> <1>
```
Comment (lihongyan1) [inline modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:50]: It's fixed, don't need substitute
writeRelabelConfigs:

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+          - <label_config> <2>
```
Comment (lihongyan1) [inline modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:51]: <relabel_config>

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+      remoteWrite:
+      - url: "https://remote-write-endpoint.example.com"
+        writeRelabelConfigs:
+        - sourceLabels: [__name__]
+          regex: 'my_metric'
+          action: keep
+        - sourceLabels:
+            - __tmp_openshift_cluster_id__ <1>
+          targetLabel: cluster_id_1 <2>
+          action: replace <3>
```
Comment (lihongyan1) [inline modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:81]: Should change as the following

```
        writeRelabelConfigs:
        - sourceLabels:
          - __tmp_openshift_cluster_id__
          targetLabel: cluster_id
```

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc (hunk 4 above)
Comment (bburt-rh) [reply modules/monitoring-creating-cluster-id-labels-for-metrics.adoc]: fixed

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc (hunk 3 above)
Comment (bburt-rh) [reply modules/monitoring-creating-cluster-id-labels-for-metrics.adoc]: fixed

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      remoteWrite:
+      - url: "https://remote-write-endpoint.example.com"
+        writeRelabelConfigs:
+        - sourceLabels: [__name__]
+          regex: 'my_metric'
+          action: keep
+        - sourceLabels:
+          - __tmp_openshift_cluster_id__ <1>
+          targetLabel: cluster_id <2>
+          action: replace <3>
```
Comment (lihongyan1) [inline modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:77]: line 71 to 73 should be deleted also, though there are valid in prometheus CRD, but not here.
line 77 action should not be in configmap, the action will be added automatically to prometheus CRD. Detail please refer to my test case or Dev's auto case
https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-49176 user/password:readonly/polarion
https://github.com/openshift/cluster-monitoring-operator/pull/1578/files#diff-b388a528a3665a46f4dfa8eb4a8c02de116acb8f85e4611f12f11dc337d7f7b5 from line 887 to 1040

Comment (jan--f) [reply modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:77]: I agree on `my_metric`. This seems confusing here. We have such a case in our e2e tests just to cover a corner case. The description adds to the confusion. In its current version two metrics are forwarded: `my_metric` and `cluster_id` but they have no relation to each other beyond being send to the remote endpoint.

On leaving out the action I'm conflicted. It is correct that `replace` is the default action in case no action is specified. I am however usually a fan of being explicit, especially in documentation. @bburt-rh maybe the doc team has a policy how default values are normally treated in config fragments like this? Maybe its worth keep the action but pointing out that its is optional since its the default value?

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+      remoteWrite:
+      - url: "https://remote-write-endpoint.example.com"
+        <endpoint_authentication_credentials>
+        writeRelabelConfigs: <1>
+          - <relabel_config> <2>
+----
+<1> Add a list of write relabel configurations for metrics that you want to send to the remote endpoint.
+<2> Substitute the label configuration for the metrics sent to the remote write endpoint.
++
+The following sample shows how to forward a single metric called `my_metric` with the cluster ID label `cluster_id_1` in default platform monitoring:
```
Comment (lihongyan1) [inline modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:56]: better change `cluster_id_1` to `cluster_id`

Comment (bburt-rh) [reply modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:56]: fixed

File: modules/monitoring-adding-cluster-id-labels-to-metrics.adoc [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+//
+// * monitoring/configuring-the-monitoring-stack.adoc
+
+:_content-type: CONCEPT
+[id="adding-cluster-id-labels-to-metrics_{context}"]
+= Adding cluster ID labels to metrics
+
+If you manage multiple {product-title} clusters and use the remote write feature to send metrics data from these clusters to an external storage location, you can add cluster ID labels to identify the metrics data coming from different clusters.
+
+You add cluster ID labels to metrics by configuring the write relabel settings for remote write storage.
```
Comment (jc-berger) [inline modules/monitoring-adding-cluster-id-labels-to-metrics.adoc:11]: There seem to be several actions in this second paragraph:

1. Configure the write relabel settings for remote write storage. 
2. Add cluster ID labels to the metrics. 
3. Query these labels to identify the source cluster for a metric and distinguish that metric data from similar metric data sent by other clusters. 

Could this content be turned into a small procedure? 

What do you think of this structure?: 

Abstract:

If you manage multiple {product-title} clusters and use the remote write feature to send metrics data from these clusters to an external storage location, you can add cluster ID labels to identify the metrics data coming from different clusters. This way, when you manage many clusters for multiple customers and send metric data to a single centralized storage system, you can use cluster ID labels to query metrics for a particular cluster or customer.

.Procedure

1. Configure the write relabel settings for remote write storage. 
2. Add cluster ID labels to the metrics. 
3. Query these labels to identify the source cluster for a metric and distinguish that metric data from similar metric data sent by other clusters. 

.Additional resources 

- Creating Cluster ID labels for metrics<link>

Comment (jc-berger) [reply modules/monitoring-adding-cluster-id-labels-to-metrics.adoc:11]: Note, I'm not saying this content _needs_ to be changed according to modularization principles. 

It simply seems to me that you have an intro/abstract followed by a series of actions, which to me, signifies a procedure module. 

What do you think?

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 9]
```diff
@@ -0,0 +1,95 @@
+// Module included in the following assemblies:
+//
+// * monitoring/configuring-the-monitoring-stack.adoc
+
+:_content-type: PROCEDURE
+[id="creating-cluster-id-labels-for-metrics_{context}"]
+= Creating cluster ID labels for metrics
+
+You can configure cluster ID labels for metrics for default platform monitoring and for user workload monitoring.
```
Comment (jc-berger) [inline modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:9]: For minimalism and user focus, can you omit, `You can` and simply say, `Configure cluster ID labels for metrics for default platform and user workload monitoring.` 

Note, if default platform and user workload have _separate_ monitoring, then keep it how you have written. If share monitoring, no need to say it twice, you can use `monitoring` once like I did.

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+For default platform monitoring, you add cluster ID labels for metrics in the `write_relabel` settings for remote write storage in the `cluster-monitoring-config` config map in the `openshift-monitoring` namespace.
```
Comment (jc-berger) [inline modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:11]: Minimalism: do you need either of these paragraphs for default platform monitoring and user workload monitoring? You very well might. 

These actions aren't included in the procedure nor in the prereqs. They don't seem necessary for the intro/abstract which is meant to set the procedure module up. 

Maybe this content would be better in a `Verification` section? So that way, users have completed the procedure on creating cluster ID labels and can verify that they can add these cluster ID labels to their monitoring workloads (default platform or user workload)? 

What do you think: keep where it is, omit it, move it to a verification section?

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        writeRelabelConfigs:
+        - sourceLabels:
+          - __tmp_openshift_cluster_id__ <1>
+          targetLabel: cluster_id <2>
+          action: replace <3>
+----
+<1> The system initially applies a temporary cluster ID source label named `+++__tmp_openshift_cluster_id__+++`. This temporary label is later replaced with the cluster ID label name that you specify.
+<2> Specify the name of the cluster ID label for metrics sent to remote write storage. 
+If you use a label name that already exists for a metric, that value is overwritten with the name of this cluster ID label.
+Do not use `+++__tmp_openshift_cluster_id__+++` for the label name. The final relabeling step removes labels that use this name.
```
Comment (jc-berger) [inline modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:79]: Personal  preference, but I think this reorder reads better: `For the label name, do not use `+++__tmp_openshift_cluster_id__+++` `

File: monitoring/configuring-the-monitoring-stack.adoc [hunk 12]
```diff
@@ -97,6 +97,15 @@ include::modules/monitoring-supported-remote-write-authentication-settings.adoc[
...
 * See the xref:../rest_api/monitoring_apis/prometheus-monitoring-coreos-com-v1.adoc#spec-remotewrite-2[Prometheus REST API reference for remote write] for information about additional optional fields.
 
+// Configuring labels for outgoing metrics
+include::modules/monitoring-adding-cluster-id-labels-to-metrics.adoc[leveloffset=+1]
+include::modules/monitoring-creating-cluster-id-labels-for-metrics.adoc[leveloffset=+2]
+
+[role="_additional-resources"]
+.Additional resources
+
+* See xref:../monitoring/configuring-the-monitoring-stack.adoc#configuring_remote_write_storage_configuring-the-monitoring-stack[Configuring remote write storage] for details about write relabel configuration.
```
Comment (jc-berger) [inline monitoring/configuring-the-monitoring-stack.adoc:107]: For scannability and usability, I find it easier to click links when they're not sandwhiched in between words. What do you think of this reorder?: 

`For details about write relabel configuration, see Configuring remote write storage<link>.`

**Review by jc-berger — COMMENTED** [review]
I used this openshift checklist when conducting this peer review: https://docs.google.com/document/d/1WN7k72PKxcPA__erp5TK-CefmS_sVwE3r40PvpmNk3E/edit#heading=h.mgz8v4w1kg86

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc (hunk 9 above)
Comment (bburt-rh) [reply modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:9]: They are separate monitoring systems, so leaving as-is.

File: modules/monitoring-adding-cluster-id-labels-to-metrics.adoc (hunk 8 above)
Comment (bburt-rh) [reply modules/monitoring-adding-cluster-id-labels-to-metrics.adoc:11]: This is a concept module, and the procedure comes next in a different file, but I like the idea of including a conceptual overview of the overall procedure here.  I also like your rewrite of the conceptual content and will use it to revise. I'll try a revision and ping you to have a look.

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    prometheusK8s:
+      remoteWrite:
+      - url: "https://remote-write-endpoint.example.com"
+        writeRelabelConfigs:
+        - sourceLabels:
+          - __tmp_openshift_cluster_id__ <1>
+          targetLabel: cluster_id <2>
+          action: replace <3>
+----
+<1> The system initially applies a temporary cluster ID source label named `+++__tmp_openshift_cluster_id__+++`. This temporary label is later replaced with the cluster ID label name that you specify.
```
[the remark this answers is not in the document]
Comment (bburt-rh) [reply modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:76]: yes, it does. Thanks. Will change

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc (hunk 11 above)
Comment (bburt-rh) [reply modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:79]: I agree. Makes better linkage to the phrasing at the end of the previous sentence. Will revise per your suggestion.

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+This action is the default value and is applied if no value is specified.
++
+[NOTE]
+====
+If you configure cluster ID labels for metrics for the Prometheus instance that monitors user-defined projects, edit the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace. 
+Note that the Prometheus component is called `prometheus` in this config map and not `prometheusK8s`, which is the name used in the `cluster-monitoring-config` config map. 
+====
+
+. Save the file to apply the changes to the `ConfigMap` object.
+The pods affected by the updated configuration restart automatically.
```
[the remark this answers is not in the document]
Comment (bburt-rh) [reply modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:90]: Agree. Fixed.

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc (hunk 10 above)
Comment (jc-berger) [reply modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:11]: Makes sense, Brian. In that case, leave as is :)

File: monitoring/configuring-the-monitoring-stack.adoc (hunk 12 above)
Comment (bburt-rh) [reply monitoring/configuring-the-monitoring-stack.adoc:107]: Agree. will make the change per your suggestion

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+          action: replace <3>
+----
+<1> The system initially applies a temporary cluster ID source label named `+++__tmp_openshift_cluster_id__+++`. This temporary label is later replaced with the cluster ID label name that you specify.
+<2> Specify the name of the cluster ID label for metrics sent to remote write storage. 
+If you use a label name that already exists for a metric, that value is overwritten with the name of this cluster ID label.
+Do not use `+++__tmp_openshift_cluster_id__+++` for the label name. The final relabeling step removes labels that use this name.
+<3> The `replace` write relabel action replaces the temporary label with the target label for outgoing metrics. 
+This action is the default value and is applied if no value is specified.
++
+[NOTE]
```
[the remark this answers is not in the document]
Comment (bburt-rh) [reply modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:83]: Moved it up after step one. It makes more sense there because that's where you issue the command to edit the config map.

File: modules/monitoring-adding-cluster-id-labels-to-metrics.adoc [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+// * monitoring/configuring-the-monitoring-stack.adoc
+
+:_content-type: CONCEPT
+[id="adding-cluster-id-labels-to-metrics_{context}"]
+= Adding cluster ID labels to metrics
+
+If you manage multiple {product-title} clusters and use the remote write feature to send metrics data from these clusters to an external storage location, you can add cluster ID labels to identify the metrics data coming from different clusters. 
+You can then query these labels to identify the source cluster for a metric and distinguish that data from similar metrics data sent by other clusters.
+
+This way, if you manage many clusters for multiple customers and send metrics data to a single centralized storage system, you can use cluster ID labels to query metrics for a particular cluster or customer.
```
Comment (jc-berger) [inline modules/monitoring-adding-cluster-id-labels-to-metrics.adoc:12]: Nice, I like that you state the benefit to users if they add cluster ID to metrics.

File: modules/monitoring-adding-cluster-id-labels-to-metrics.adoc [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+
+Creating and using cluster ID labels involves three general steps:
```
Comment (jc-berger) [inline modules/monitoring-adding-cluster-id-labels-to-metrics.adoc:14]: Good, I think add these three general steps to the concept module helps set the user up for success as they proceed with the procedures.

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+.Procedure
+
+. Edit the `cluster-monitoring-config` `ConfigMap` object in the `openshift-monitoring` project:
++
+[source,terminal]
+----
+$ oc -n openshift-monitoring edit configmap cluster-monitoring-config
+----
++
+[NOTE]
```
Comment (jc-berger) [inline modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:35]: Good, I think the Note reads well here.

**Review by jc-berger — APPROVED** [review]

File: modules/monitoring-adding-cluster-id-labels-to-metrics.adoc (hunk 16 above)
Comment (bburt-rh) [reply modules/monitoring-adding-cluster-id-labels-to-metrics.adoc:12]: thank, Jake.

File: modules/monitoring-adding-cluster-id-labels-to-metrics.adoc (hunk 17 above)
Comment (bburt-rh) [reply modules/monitoring-adding-cluster-id-labels-to-metrics.adoc:14]: That sets my mind at ease about it. Thx

**Review by jc-berger — APPROVED** [review]
/lgtm

File: modules/monitoring-creating-cluster-id-labels-for-metrics.adoc [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+          - __tmp_openshift_cluster_id__ <1>
+          targetLabel: cluster_id <2>
+          action: replace <3>
+----
+<1> The system initially applies a temporary cluster ID source label named `+++__tmp_openshift_cluster_id__+++`. This temporary label gets replaced by the cluster ID label name that you specify.
+<2> Specify the name of the cluster ID label for metrics sent to remote write storage. 
+If you use a label name that already exists for a metric, that value is overwritten with the name of this cluster ID label.
+For the label name, do not use `+++__tmp_openshift_cluster_id__+++`. The final relabeling step removes labels that use this name.
+<3> The `replace` write relabel action replaces the temporary label with the target label for outgoing metrics. 
+This action is the default value and is applied if no value is specified.
```
Comment (jan--f) [inline modules/monitoring-creating-cluster-id-labels-for-metrics.adoc:87]:
```suggestion
This action is the default value and is applied if no action is specified.
```
One last nitpick, since the value of of `action` can't be empty and would be rejected. `action` can be absent however and in that case the default comes into play. Otherwise the looks great, thanks!

**Review by jan--f — CHANGES_REQUESTED** [review]
